---
float32[] pos
int32[] enc
float32[] ref
